server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -87,6 +86,64 @@ public Builder add(Mapper.Builder builder) {
             return this;
         }
 
+        Builder addMappers(Map<String, Mapper> mappers) {
+            mappers.forEach((name, mapper) -> mappersBuilders.add(new Mapper.Builder(name) {
+                @Override
+                public Mapper build(MapperBuilderContext context) {
+                    return mapper;
+                }
+            }));
+            return this;
+        }
+
+        /**
+         * Adds a dynamically created Mapper to this builder.
+         *
+         * @param name      the name of the Mapper, including object prefixes
+         * @param prefix    the object prefix of this mapper
+         * @param mapper    the mapper to add
+         * @param context   the DocumentParserContext in which the mapper has been built
+         */
+        public void addDynamic(String name, String prefix, Mapper mapper, DocumentParserContext context) {
+            // If the mapper to add has no dots and is therefore
+            // a leaf mapper, we just add it here
+            if (name.contains(".") == false) {
+                mappersBuilders.add(new Mapper.Builder(name) {
+                    @Override
+                    public Mapper build(MapperBuilderContext context) {
+                        return mapper;
+                    }
+                });
+            }
+            // otherwise we strip off the first object path of the mapper name, load or create
+            // the relevant object mapper, and then recurse down into it, passing the remainder
+            // of the mapper name. So for a mapper 'foo.bar.baz', we locate 'foo' and then
+            // call addDynamic on it with the name 'bar.baz'.
+            else {
+                int firstDotIndex = name.indexOf(".");
+                String childName = name.substring(0, firstDotIndex);
+                String fullChildName = prefix == null ? childName : prefix + "." + childName;
+                ObjectMapper.Builder childBuilder = findChild(childName, fullChildName, context);
+                childBuilder.addDynamic(name.substring(firstDotIndex + 1), fullChildName, mapper, context);
+                mappersBuilders.add(childBuilder);
+            }
+        }
+
+        private ObjectMapper.Builder findChild(String childName, String fullChildName, DocumentParserContext context) {
+            // does the child mapper already exist? if so, use that
+            ObjectMapper child = context.mappingLookup().objectMappers().get(fullChildName);
```

Comment:
These top two calls only resolve stuff on the first iteration, right? Once you recur you'll know this never resolves?

Reply:
Let's say we have an existing mapping of `foo.bar`, and we get an input document with `foo.bar.one.two`.  Document parsing will add two new mappers for us, a new object `foo.bar.one` and a leaf mapper `foo.bar.one.two`.  (The reason we add the object mapper as we go is that we use dynamic templates to do this, and so we might want to create a nested mapper here, or find that it's a disabled object mapper and so skip adding the leaf entirely).

To build our dynamic update, we start with an empty root.  We call `addDynamic()` on this root with `foo.bar.one.two` as a mapper (intermediate objects don't get passed to addDynamic).  It passes `foo` as the fullName to `findChild()`, and `foo` indeed already exists, so we pull a new empty build from it, return that, add it to our empty root builder, and then recurse into it, chopping `foo` off the path.

Now we're in `addDynamic()` on `foo`, with a name of `bar.one.two` and a prefix of `foo`.  We call `findChild()` with `bar` as the childName and `foo.bar` as the fullChildName.  And indeed, `foo.bar` already exists in the mappings, so we pull a new empty builder from it, and return that; the new builder gets added to our otherwise empty `foo.bar` builder, and we recurse again.

So we're in `addDynamic()` on `foo.bar`, with a name of `one.two` and a prefix of `foo.bar`.  We call `findChild` for `one`, which first tries to lookup `foo.bar.one` in the mappings, and lo! It doesn't exist.  So we move to the next step, do we have a `foo.bar.one` in the newly created dynamic object mappings.  And it *does* exist there.  So we pull an empty builder from it and return that.  And we recurse again.

Now we're on a leaf - our name is `two`, which contains no dots, and so we can just add it directly to the builder and return back up the stack.

Let's say we also have another new mapper, `foo.bar.one.three`; this will follow the same path, but actually builds an entirely new tree of builders and adds that to the root.  So once that is done, our root builder contains *two* child builders called `foo`, each of which contains a single builder called `bar`, each of which contains a single builder called `one`, each of which contains a single leaf called either `two` or `three`.  When we call `build()` on the root builder, everything gets merged together.